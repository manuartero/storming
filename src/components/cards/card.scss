@use '/src/styles/fonts.scss';
@use '/src/styles/shadows.scss';
@use '/src/styles/colors.scss';

$card-text-content-width: 120px;

@mixin card-icon($action, $owner) {
  background-image: url('./assets/' + $action + '-icn--' + $owner + '.svg');
  background-position: center;
  background-repeat: no-repeat;
  background-size: contain;
}

@mixin card-water-mark($action) {
  background-image: url('./assets/' + $action + '-bg.svg');
  background-position: center;
  background-repeat: no-repeat;
  background-size: contain;
}

@mixin card-container() {
  width: 180px;
  height: 280px;
  padding: 36px 24px;
  background: #FFFFFF;
  margin-top: 2vh;
  margin-bottom: 2vh;
  @include shadows.simple-box-shadow();

}

.card {
  @include card-container();

  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: left;

  // border: 3px none transparent;

  &--played {

    // to every children
    &>* {
      opacity: 0.3;
    }
  }

  &--selected {
    background-color: colors.$selected-grey;
  }

  .action-card {

    &__icon {
      width: 70px;
      height: 55px;

      &--build-player {
        @include card-icon('build', 'player')
      }

      &--build-enemy1 {
        @include card-icon('build', 'enemy1')
      }

      &--build-enemy2 {
        @include card-icon('build', 'enemy2')
      }

      &--build-enemy3 {
        @include card-icon('build', 'enemy3')
      }

      &--diplo-player {
        @include card-icon('diplo', 'player')
      }

      &--diplo-enemy1 {
        @include card-icon('diplo', 'enemy1')
      }

      &--diplo-enemy2 {
        @include card-icon('diplo', 'enemy2')
      }

      &--diplo-enemy3 {
        @include card-icon('diplo', 'enemy3')
      }

      &--move-player {
        @include card-icon('move', 'player')
      }

      &--move-enemy1 {
        @include card-icon('move', 'enemy1')
      }

      &--move-enemy2 {
        @include card-icon('move', 'enemy2')
      }

      &--move-enemy3 {
        @include card-icon('move', 'enemy3')
      }

      &--recruit-player {
        @include card-icon('recruit', 'player')
      }

      &--recruit-enemy1 {
        @include card-icon('recruit', 'enemy1')
      }

      &--recruit-enemy2 {
        @include card-icon('recruit', 'enemy2')
      }

      &--recruit-enemy3 {
        @include card-icon('recruit', 'enemy3')
      }
    }

    &__title {
      @include fonts.title;

      width: $card-text-content-width;
      height: 24px;

      text-transform: uppercase;
    }

    &__text {
      @include fonts.paragraph;

      width: $card-text-content-width;
      height: 90px;
    }

    &__water-mark {
      position: absolute;
      bottom: 0;
      right: 0;
      width: 60%;
      height: 60%;

      &--build {
        @include card-water-mark('build');
      }

      &--diplo {
        @include card-water-mark('diplo');
      }

      &--move {
        @include card-water-mark('move');
      }

      &--recruit {
        @include card-water-mark('recruit');
      }
    }
  }

}
