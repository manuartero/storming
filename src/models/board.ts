import { tilesInRange } from "./tiles";

type _FilterPredicate = (_: [string, Tile]) => boolean;

function hasBuildingFromSameOwner(card: ActionCard): _FilterPredicate {
  return ([_, tile]) => tile.building?.owner === card.owner;
}

function hasPieceFromSameOwner(card: ActionCard): _FilterPredicate {
  return ([_, tile]) => tile.piece?.owner === card.owner;
}

function hasSoldierFromSameOwner(card: ActionCard): _FilterPredicate {
  return ([_, tile]) =>
    tile.piece?.owner === card.owner && tile.piece?.type === "soldier";
}

function asTileID([t, _]: [string, Tile]): TileID {
  return t as TileID;
}

export function getAvailableTilesForActionCard({
  board,
  activeCard,
  selectedTile,
  activePlayer,
}: {
  board: Board;
  activeCard: ActionCard;
  selectedTile: TileID | undefined;
  activePlayer: Player | undefined;
}): TileID[] {
  if (activeCard.action === "move") {
    if (selectedTile && board[selectedTile].piece?.owner === activePlayer) {
      return getInRangeMovements(board, selectedTile);
    }
    return Object.entries(board)
      .filter(hasPieceFromSameOwner(activeCard))
      .map(asTileID);
  }

  if (activeCard.action === "build") {
    return Object.entries(board)
      .filter(hasSoldierFromSameOwner(activeCard))
      .map(asTileID);
  }

  if (activeCard.action === "recruit") {
    return Object.entries(board)
      .filter(hasBuildingFromSameOwner(activeCard))
      .map(asTileID);
  }
  return [];
}

export function getInRangeMovements(board: Board, tile: TileID): TileID[] {
  const tiles = tilesInRange(tile);
  return tiles.filter(
    (candidateTile) =>
      /* valid options: */
      !board[candidateTile].piece ||
      board[candidateTile].piece?.owner !== board[tile].piece?.owner
  );
}
