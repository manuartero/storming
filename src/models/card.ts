function cardModel(card: Card | undefined) {
  return {
    toString() {
      if (!card) {
        return "undefined";
      }
      return card.cardType === "actionCard" ? card.action : card.event;
    },
  };
}

export default cardModel;
