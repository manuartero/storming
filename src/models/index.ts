export { Card } from "./card";
export { PLAYER_CARDS } from "./player-cards";
