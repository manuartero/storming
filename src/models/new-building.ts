export function NewBuilding({ owner }: { owner: PlayerType }): Building {
  const building = {
    owner,
    type: "tower" as const,
    hasWalls: false,
  };
  const toString = () => `Building{ ${building.type}, ${building.owner} }`;
  Object.setPrototypeOf(building, { toString });
  return building;
}

export function upgradeBuilding(building: Building): Building {
  if (building.type === "tower") {
    return {
      ...building,
      type: "castle",
    };
  }
  if (building.type === "castle") {
    return {
      ...building,
      type: "citadel",
    };
  }
  return building;
}
