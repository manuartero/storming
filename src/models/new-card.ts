const cardIdCount = new Map<string, number>();

export function _resetCardId() {
  cardIdCount.clear();
}

export function isActionCard(card: Card): card is ActionCard {
  return card.cardType === "actionCard";
}

export function isEventCard(card: Card): card is EventCard {
  return card.cardType === "eventCard";
}

function isActionCardType(
  type: ActionCardType | EventCardType
): type is ActionCardType {
  return (
    type === "build" ||
    type === "diplo" ||
    type === "move" ||
    type === "recruit"
  );
}

function getCardId(
  type: ActionCardType | EventCardType,
  player: PlayerType
): CardId {
  const baseCardId = `${player}_${type}` as const;
  const count = cardIdCount.get(baseCardId) || 1;
  cardIdCount.set(baseCardId, count + 1);
  const cardId = `${baseCardId}_${count}` as const;
  return cardId;
}

export function NewCard(type: ActionCardType, player: PlayerType): ActionCard;

export function NewCard(type: EventCardType, player: PlayerType): EventCard;

export function NewCard(
  type: ActionCardType | EventCardType,
  player: PlayerType
): ActionCard | EventCard {
  const cardId = getCardId(type, player);
  const toString = () => `Card{ ${type} (${player}) }`;

  const card = isActionCardType(type)
    ? {
        cardType: "actionCard" as const,
        action: type,
        owner: player,
        cardId,
      }
    : {
        cardType: "eventCard" as const,
        event: type,
        playedBy: player,
        cardId,
      };

  Object.setPrototypeOf(card, { toString });
  return card;
}
