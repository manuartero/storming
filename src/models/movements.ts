import { asTileID, coordinates } from "models/tiles";

export function getAvailableMovements(t: TileID, range = 1) {
  console.debug(`getAvailableMovements(${t})`);
  const tileID = coordinates(t);

  // TODO use range param instead of fixed array
  // TODO avoid fixed arrays?
  const range1Movements =
    tileID.y % 2 === 0
      ? [
          [0, -1],
          [+1, -1],
          [-1, 0],
          [+1, 0],
          [0, +1],
          [+1, +1],
        ]
      : [
          [-1, -1],
          [0, -1],
          [-1, 0],
          [+1, 0],
          [-1, +1],
          [0, +1],
        ];

  return range1Movements.flatMap((variance) => {
    const tileIdStr = asTileID({
      x: tileID.x + variance[0],
      y: tileID.y + variance[1],
    });
    return tileIdStr ? [tileIdStr] : [];
  });
}
