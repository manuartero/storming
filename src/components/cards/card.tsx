import c from "classnames";
import { isActionCard, isEventCard } from "models/new-card";
import CARD_TEXT from "./card-text.json";
import { actionCardAssets } from "./assets";

import styles from "./card.module.css";
import fontStyles from "styles/fonts.module.css";

type Props = {
  card: Card;
  status?: PlayerHandCardStatus | CardStatus;
  onClick?: () => void;
};

export function Card({ card, status = "available", onClick }: Props) {
  return (
    <div
      className={c(
        styles.card,
        styles[status],
        isActionCard(card) && styles.actionCard,
        isEventCard(card) && styles.eventCard,
        onClick && status === "available" && styles.clickable
      )}
      key={card.cardId}
      aria-label={`card ${card.cardId}`}
      aria-pressed={status === "selected"}
      aria-disabled={!onClick}
      onClick={onClick}
    >
      {card.cardType === "actionCard" && <ActionCardContents card={card} />}
      {card.cardType === "eventCard" && <EventCardContents card={card} />}
    </div>
  );
}

function ActionCardContents({ card: actionCard }: { card: ActionCard }) {
  const { action, owner } = actionCard;
  const cardIcon = actionCardAssets[action][owner];
  const backgroundWaterMark = actionCardAssets[action].bgWaterMark;

  return (
    <>
      <div
        className={styles.actionCardIcon}
        style={{ backgroundImage: `url(${cardIcon})` }}
      ></div>
      <div className={c(styles.actionCardTitle, fontStyles.title)}>
        {action}
      </div>
      <div className={c(styles.actionCardText, fontStyles.paragraph)}>
        {CARD_TEXT[action]}
      </div>
      <div
        className={styles.actionCardWaterMark}
        style={{ backgroundImage: `url(${backgroundWaterMark})` }}
      ></div>
    </>
  );
}

function EventCardContents({ card }: { card: EventCard }) {
  // TODO: Alpha
  return <div></div>;
}
