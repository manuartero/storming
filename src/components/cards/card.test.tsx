import { render, screen } from "@testing-library/react";
import { NewCard } from "models/new-card";
import { Card } from "./card";

const actionCard = NewCard("build", "player");

describe("<Card />", () => {
  it("render: article role", () => {
    render(<Card card={actionCard} onClick={jest.fn()} />);

    const card = screen.getByRole("article", {
      name: `card ${actionCard.cardId}`,
    });
    expect(card).toMatchSnapshot();
  });
});
