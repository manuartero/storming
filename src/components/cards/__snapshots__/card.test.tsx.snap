// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Card /> render: article role 1`] = `
<article
  aria-disabled="false"
  aria-label="card player_build_1"
  aria-pressed="false"
  class="card available actionCard"
>
  <div
    class="heading"
  >
    <div
      class="icon"
    />
    <div
      class="title title"
    >
      build
    </div>
  </div>
  <div
    class="content"
  >
    <div
      class="text paragraph"
    >
      <p>
        Choose one:
      </p>
      <p>
        Found a new Settlement
      </p>
      <p>
        Improve a City
      </p>
      <p>
        Build Walls on a Settlement
      </p>
    </div>
    <div
      class="waterMark"
    />
  </div>
</article>
`;
